### PR TITLE
Fix writing BLOBs to postgres with recent contacts interaction

### LINKED ISSUE #22330: postgresql - error while creating share (invalid input syntax for type bytea)
### Steps to reproduce
1. NC installation with LDAP users that have a "," (comma) in their displayname.
2. choose any file in the root of your own nextcloud directory
3. click on the "share" symbol
4. search for an LDAP user and choose him

### Expected behaviour
File should be shared without error message and should appear in the files view of the choosen user.

### Actual behaviour
Error message "error on sharing file" appears. Choosen user ist not listed in share list. After refreshing the browser the user is listed but the file doesn't appear in the files view of the choosen user.
The browser logs says: "Invalid text representation: 7 ERROR: invalid input syntax for type bytea"

### Server configuration

**Operating system:**
Ubuntu 18.04.4 LTS

**Web server:**
Apache2 2.4.29

**Database:**
postgresql 10.12

**PHP version:**
php 7.2.24

**Nextcloud version:**
19.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated from 18.x

**Where did you install Nextcloud from:**
Stable release channel

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forwarded_for_headers": [
            "HTTP_X_FORWARDED",
            "HTTP_FORWARDED_FOR"
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "memcache.local": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": "true",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0,
            "dbindex": 0
        },
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "overwriteprotocol": "https",
        "dbtype": "pgsql",
        "version": "19.0.1.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "ldapUserCleanupInterval": "20",
        "sharing.maxAutocompleteResults": 0,
        "sharing.minSearchStringLength": 0,
        "sharing.enable_share_accept": false,
        "sharing.force_share_accept": false,
        "sharing.enable_share_mail": true,
        "updater.release.channel": "stable",
        "onlyoffice": {
            "verify_peer_off": true
        },
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "htaccess.RewriteBase": "\/",
        "maintenance": false,
        "theme": "",
        "log_type": "file",
        "loglevel": 1,
        "auth.bruteforce.protection.enabled": true,
        "app_install_overwrite": [
            "sharepoint"
        ],
        "mail_sendmailmode": "smtp"
    }
}

```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/ftp

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP

### Client configuration
**Browser:** Firefox

**Operating system:** Windows 10

#### Nextcloud log (data/nextcloud.log)
no entries regarding this issue

#### Browser log
<details>
<summary>Browser log</summary>

```
POSThttps://nextcloud.tld/ocs/v2.php/apps/files_sharing/api/v1/shares [HTTP/1.1 403 Forbidden 1112ms]

XHRPOSThttps://nextcloud.tld/ocs/v2.php/apps/files_sharing/api/v1/shares
[HTTP/1.1 403 Forbidden 1112ms]

    	
    ocs	Object { meta: {…}, data: [] }
    meta	Object { status: "failure", statuscode: 403, message: "An exception occurred while executing 'INSERT INTO \"oc_recent_contact\" (\"actor_uid\", \"uid\", \"last_contact\", \"card\") VALUES(?, ?, ?, ?)' with params [\"x.x\", \"y.y\", 1597922779, \"BEGIN:VCARD\\r\\nVERSION:4.0\\r\\nPRODID:-\\/\\/Sabre\\/\\/Sabre VObject 4.3.0\\/\\/EN\\r\\nUID:sabre-vobject-d4ce47f0-be9d-474a-992d-9350efdafbcf\\r\\nURI:2bd9efd6-c866-4ef5-c6c5-c3ce3e5c3869\\r\\nFN:y\\\\, y\\r\\nCATEGORIES:Recently contacted\\r\\nX-NEXTCLOUD-UID:y.y\\r\\nEND:VCARD\\r\\n\"]:\n\nSQLSTATE[22P02]: Invalid text representation: 7 ERROR: invalid input syntax for type bytea" }
    status	"failure"
    statuscode	403
    message	"An exception occurred while executing 'INSERT INTO \"oc_recent_contact\" (\"actor_uid\", \"uid\", \"last_contact\", \"card\") VALUES(?, ?, ?, ?)' with params [\"x.x\", \"y.y\", 1597922779, \"BEGIN:VCARD\\r\\nVERSION:4.0\\r\\nPRODID:-\\/\\/Sabre\\/\\/Sabre VObject 4.3.0\\/\\/EN\\r\\nUID:sabre-vobject-d4ce47f0-be9d-474a-992d-9350efdafbcf\\r\\nURI:2bd9efd6-c866-4ef5-c6c5-c3ce3e5c3869\\r\\nFN:y\\\\, y\\r\\nCATEGORIES:Recently contacted\\r\\nX-NEXTCLOUD-UID:y.y\\r\\nEND:VCARD\\r\\n\"]:\n\nSQLSTATE[22P02]: Invalid text representation: 7 ERROR: invalid input syntax for type bytea"
    data	[]

```
</details>

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/server/issues/22330

### CODE REVIEW

File: lib/public/AppFramework/Db/Entity.php [hunk 1]
```diff
@@ -110,7 +110,12 @@ protected function setter($name, $args) {
 
 			// if type definition exists, cast to correct type
 			if ($args[0] !== null && array_key_exists($name, $this->_fieldTypes)) {
-				settype($args[0], $this->_fieldTypes[$name]);
+				$type = $this->_fieldTypes[$name];
+				if ($type === 'blob') {
+					// (B)LOB is treated as string when we read from the DB
+					$type = 'string';
+				}
```
Comment (blizzz) [inline lib/public/AppFramework/Db/Entity.php:117]: is it  safe for all scenarios?

Comment (ChristophWurst) [reply lib/public/AppFramework/Db/Entity.php:117]: @nickvergessen made me do this :speak_no_evil:

Comment (ChristophWurst) [reply lib/public/AppFramework/Db/Entity.php:117]: AFAIK it won't be any worse than what we have right now

**Review by nickvergessen — APPROVED** [review]

File: lib/public/AppFramework/Db/Entity.php (hunk 1 above)
Comment (nickvergessen) [reply lib/public/AppFramework/Db/Entity.php:117]: I think it should be save. Because even binary data is a string in php

**Review by blizzz — APPROVED** [review]

**Comment by tcitworld** [comment]
This doesn't seem to work properly.
PostgreSQL indeed returns a resource when using `PARAM_LOB`, not a string. Converting it like this doesn't read the stream and the string returned is always something like `Resource id #14`.

In the case of `contactsinteraction` it then fails when we try to read VCS data, leading to https://github.com/nextcloud/server/issues/22439

**Pull request merged**